Исаев Сергей

Luvr681 opened the pull request:
Решение на проверку

@dmitgaranin

Luvr681 committed:
Решение на проверку

dmitgaranin commented on ErrorHandling/ErrorHandling.csproj:
```diff
@@ -2,7 +2,7 @@
 
     <PropertyGroup>
```
Сразу о ведении гита)

1. Разбивай изменения на единицы изменений. В твоем (единственном...) коммите лежит все-все-все решение. Так делать точно не стоит, т.к.:
- по истории коммитов не понятно, что именно происходило в решении
- если придется откатывать какой-то кусок кода, то это будет невозможно
- иногда бывает полезно переносить коммиты по веткам, но вряд ли будет нужно переносить все-все-все

2. Давай коммитам более осмысленные имена, чем `Решение на проверку` :) Как минимум, в данном случае не очень понятно, что за решение, почему контекст о проверке лежит на гите, а не на МР, ну и все вытекающее. Тут думаю, можно без пояснений, зачем это полезно)

Luvr681 replied:
Хорошо, понял

dmitgaranin commented on FileSenderRailway/Dependencies.cs:
```diff
@@ -1,5 +1,8 @@
 using System;
+using System.Collections;
+using System.Collections.Generic;
```
А зачем изменял проекты, не относящиеся к задаче?)

Luvr681 replied:
У нас было занятие по fp, где мы работали в этом репозитории. Я работал в форке. Так лучше не делать?

dmitgaranin replied:
Конечно) Текущая задача о конкретном ТЗ, а твои изменения содержат куда большее

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Additions/EmptyMain.cs:
```diff
@@ -0,0 +1,6 @@
+
+public class EmptyMain
+{
+    public static void Main() {} // Здесь аналогично, как и в Settings. Не компилится без Main
```
Попробуй сделать этот csproj не исполняемым приложением, а библиотекой. 
Оставлять ненужный `main`, чтобы что-то компилилось - неправильно, стоит дожимать решение без костылей)

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Additions/Models/Colors.cs:
```diff
@@ -0,0 +1,28 @@
+        if (string.IsNullOrEmpty(colorName))
+            return _colors["Black"];
+        if (!_colors.ContainsKey(colorName))
+            throw new ArgumentException("Invalid color Name");
```
Давай все-все `exceptions` заменим на `Result`

> Вместо исключений всюду в программе используй паттерн Result<T>

dmitgaranin replied:
Исключений, почему-то, стало гораздо больше. Давай все-таки уберем их все из проекта, т.к. это часть обязательного ТЗ задачи

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Additions/Validator.cs:
```diff
@@ -0,0 +1,34 @@
+
+namespace TagCloudContainer.Additions;
+
+public class Validator
```
В этом классе валидируются разные ответственности, стоит их разделить на разные классы

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Tests/Register.cs:
```diff
@@ -0,0 +1,25 @@
+
+namespace TagCloudContainer.Tests;
+
+public static class Register
```
Не нашел использований

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Tests/Validator_Should.cs:
```diff
@@ -0,0 +1,117 @@
+    public void ValidateContainerConfig_EmptyCenterPoint_ShouldBeFail()
+    {
+        var validator = new Validator();
+        var tagCloudContainerConfig = new TagCloudContainerConfig();
```
Можно использовать `object initializer`

dmitgaranin commented on TagCloudContainer/TagCloudContainer/Providers/TagCloudProvider.cs:
```diff
@@ -0,0 +1,32 @@
+    public IEnumerable<Additions.Models.Word> GetPreparedWords()
+    {
+        var sortedWords = _tagCloudContainerConfig.Random 
+            ? WordsShuffler.ShuffleWords(_words.ToList()) : _words;
```
В целом, я стараюсь не обращать внимание на моменты/недочеты с предыдущих этапов задачи, но здесь уж слишком бросается в глаза:
> Запрещено использовать операторы if, switch, ?: и прочие условные операторы, если их можно заменить полиморфизмом.

Править не предлагаю, вдруг самому будет интересно

dmitgaranin commented on TagCloudContainer/TagCloudContainer/Utils/WordsShuffler.cs:
```diff
@@ -0,0 +1,20 @@
+    public static List<TagCloudContainer.Additions.Models.Word> ShuffleWords(List<TagCloudContainer.Additions.Models.Word> words)
+    {
+        if (words == null)
+            throw new ArgumentException("Argument is null");
```
Тоже на result

dmitgaranin commented on TagCloudContainer/TagCloudContainer/Utils/TagCloudPlacer.cs:
```diff
@@ -0,0 +1,108 @@
+            }
+        }
+        
+        throw new Exception("Can't put this rectangle");
```
Тоже result

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Additions/Configs/TagCloudFormConfig.cs:
```diff
@@ -0,0 +1,13 @@
+    public string FontFamily { get; set; } = "Arial";
+    public Color Color { get; set; } = Colors.GetAll().First().Value;
+    public Color BackgroundColor { get; set; } = Colors.GetAll().First().Value;
+    public Size FormSize { get; set; } = new Size(int.Parse(Screens.Sizes.First().Split("x")[0]), 
```
Такие парсинги стоит тоже уносить за `result`

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Forms/Settings.cs:
```diff
@@ -0,0 +1,91 @@
+        _tagCloud.ShowDialog(this);
+
+        if (!_tagCloudFormResult.FormResult.IsSuccess)
+            MessageBox.Show("Облако тегов не влезло на изображение заданного размера", "Ошибка");
```
Думаю, такие вещи тоже стоит обрабатывать через `result`, а уже вне - сводить к какому-либо отображению

dmitgaranin commented on TagCloudContainer/TagCloudContainer/Utils/ImageCreator.cs:
```diff
@@ -0,0 +1,26 @@
+            {
+                g.CopyFromScreen(new Point(form.DesktopLocation.X, form.DesktopLocation.Y), new Point(0, 0), _tagCloudFormConfig.FormSize);
+            }
+            bitmap.Save(path, ImageFormat.Png);
```
А если такого пути не существует?

Luvr681 committed:
Добавил Тесты на основную валидацию конфигурационных данных

Luvr681 committed:
Перенёс регистрацию и Main в проект с формами

Luvr681 committed:
Добавил валидаторы конфигурационных данных

Luvr681 committed:
Перенёс создание итогового изображения в проект с формами

Luvr681 committed:
Выделил интерфейс валидатора

Luvr681 committed:
1. Выделил необходимые утилиты
2. Переименовал Shuffler в Arranger и избавился от тернарного оператора. Использовал полиморфизм

Luvr681 committed:
Выделил необходимые модели

Luvr681 committed:
Выделил необходимые интерфейсы

Luvr681 committed:
Оставил один конфигуратор, которые наследует разные интерфейсы для разграничения и частичного использования

Luvr681 committed:
Выделил основную логику по разположению в отдельный проект. Сделал его библиотекой

Luvr681 committed:
Выделил проект с формами

Luvr681 committed:
Удалил старый проект, поскольку перенёс новый из предыдущей задачи.
Сделал так, поскольку вносил правки по предыдущей задаче и нужно было, чтобы эти правки были и здесь

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Core/Configs/TagCloudContainerConfig.cs:
```diff
@@ -0,0 +1,43 @@
+
+namespace TagCloudContainer.Configs;
+
+public class TagCloudContainerConfig : ITagCloudContainerConfig, ITagCloudFormConfig
```
Так понимаю, в этом классе разные ответственности:
- параметры, которые из вне задает пользователь (слова исключения, цвета и тд)
- зачем-то здесь также лежит все, что связано с процессом генерации и результатом (PutRectangles, NearestToTheCenterPoints и тд)
- а еще есть методы - методов в файлах конфигурации быть не должно. Их или сделать приватным, если нужны только для инициализации дефолтных значений, или вынести куда-то вне

Сразу можно заметить, как минимум по реализации двух интерфейсов

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Core/Configs/TagCloudContainerConfig.cs:
```diff
@@ -0,0 +1,43 @@
+    public bool NeedValidate { get; set; } = true;
+
+    public string FontFamily { get; set; } = "Arial";
+    public Color Color { get; set; } = Colors.GetAll().First().Value;
```
Дублируется `Colors.GetAll().First().Value;`

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Core/Configs/TagCloudContainerConfig.cs:
```diff
@@ -0,0 +1,43 @@
+        ExcludeWordsFilePath = Path.Combine(GetMainDirectoryPath(), fileName);
+    }
+    public static string GetMainDirectoryPath() 
+        => Directory.GetParent(Directory.GetCurrentDirectory()).Parent.Parent.Parent.FullName;
```
Rider подсвечивает потенциальную ошибку 
![image](https://user-images.githubusercontent.com/8987952/209505013-f90d47e9-c362-4a0d-8a36-748257fd4cd0.png)

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Core/Configs/TagCloudContainerConfig.cs:
```diff
@@ -0,0 +1,43 @@
+
+public class TagCloudContainerConfig : ITagCloudContainerConfig, ITagCloudFormConfig
+{
+    public bool Random { get; set; } = true;
```
Случайность чего? Название не отражает суть

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Core/Configs/TagCloudContainerConfig.cs:
```diff
@@ -0,0 +1,43 @@
+    public Size StandartSize { get; set; } = new Size(10, 10);
+    public Size ImageSize { get; set; } = Screens.Sizes.First();
+    
+    public Point Center { get; set; } = new Point(1, 1);
```
Центр чего? Название не отражает суть

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Core/Configs/TagCloudContainerConfig.cs:
```diff
@@ -0,0 +1,43 @@
+    
+    public SortedList<float, Point> NearestToTheCenterPoints { get; set; } = new SortedList<float, Point>();
+    
+    public List<Rectangle> PutRectangles { get; set; } = new List<Rectangle>();
```
Только методы стоит называть через глаголы

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Core/Configs/TagCloudContainerConfig.cs:
```diff
@@ -0,0 +1,43 @@
+
+    public string ImageName { get; set; } = "TagCloudResult.png";
+    
+    public bool NeedValidate { get; set; } = true;
```
Валидировать что?

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Core/Interfaces/ISizeInvestigator.cs:
```diff
@@ -0,0 +1,8 @@
+
+public interface ISizeInvestigator
+{
+    public bool DidFit(Word word);
```
Название не очень хорошо отражает суть. К тому же стоит избегать:
- слова `Do`. Всегда можно найти глагол, подходящий лучше
- прошедшего времени в именовании методов

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Core/Interfaces/ITagCloudProvider.cs:
```diff
@@ -0,0 +1,8 @@
+
+namespace TagCloudContainer.Core.Interfaces;
+
+public interface ITagCloudProvider
```
Не очень понимаю, почему `ITagCloudProvider` что-то знает и делает в подготовке слов. Думаю, стоит переименовать

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Core/Models/Colors.cs:
```diff
@@ -0,0 +1,28 @@
+﻿namespace TagCloudContainer.Core.Models;
```
У всех файлов проблемы с namespace
![image](https://user-images.githubusercontent.com/8987952/209506459-d581addb-0785-435c-9ee7-a5987ae585ee.png)

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Core/Models/RotateDirections.cs:
```diff
@@ -0,0 +1,6 @@
+
+public enum RotateDirections
+{
+    none, bottom, left
```
IDE подсказывает, какие правила code-style для enum
![image](https://user-images.githubusercontent.com/8987952/209506577-ab757aaf-fdee-40a4-a517-6d74e7a74b16.png)

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Core/Models/Screens.cs:
```diff
@@ -0,0 +1,18 @@
+
+public static class Screens
+{
+    private static readonly Size[] _sizes = 
```
Можно сразу возвращать
```
public static IEnumerable<Size> Sizes { get; } =  new[]
    {
        new Size(1920, 1080),
        new Size(1600, 900),
        new Size(1400, 1050),
        new Size(1024, 768),
        new Size(600, 600),
    };
```

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Core/TagCloudProvider.cs:
```diff
@@ -0,0 +1,46 @@
+        ITagCloudContainerConfig tagCloudContainerConfig)
+    {
+        if (wordsReader == null)
+            throw new ArgumentNullException("Word reader can't be null");
```
Ты используешь неподходящий конструктор этой ошибки `public ArgumentNullException(string? paramName)`, т.е. сейчас в качестве имени параметра, который оказался null является не `wordsReader`, а `"Word reader can't be null"`

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Core/TagCloudProvider.cs:
```diff
@@ -0,0 +1,46 @@
+
+    public Result<List<Word>> GetPreparedWords()
+    {
+        var result = (new List<Word>()).AsResult();
```
Зачем конвертировать в Result?

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Core/TagCloudProvider.cs:
```diff
@@ -0,0 +1,46 @@
+        _words = wordsReader.GetWordsFromFile(wordsFilePath);
+    }
+
+    public Result<List<Word>> GetPreparedWords()
```
Можно вернуть `IEnumerable`

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Core/TagCloudProvider.cs:
```diff
@@ -0,0 +1,46 @@
+        var result = (new List<Word>()).AsResult();
+        var arrangedWords = WordsArranger.ArrangeWords(_words.ToList(), _tagCloudContainerConfig.Random);
+
+        if (!arrangedWords.IsSuccess)
```
Можно воспользоваться знаниями из предыдущего блока и чейнить результаты через Fluent
```
return arrangedWords
            .Then(words =>
            {
                var result = new List<Word>();
                foreach (var wordResult in words.Select(word => _tagCloudPlacer.PlaceInCloud(word)))
                {
                    if (!wordResult.IsSuccess)
                        return Result.Fail<IEnumerable<Word>>(wordResult.Error);
                    result.Add(wordResult.Value);
                }

                return result;
            });
```

По факту, и внутренний `Result` можно заменить на `Then`, но текущая реализация этого паттерна негибкая(

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Core/Utils/ImageCreator.cs:
```diff
@@ -0,0 +1,30 @@
+    
+    public void Save(Form form, string path)
+    {
+        using (Bitmap bitmap = new Bitmap(_tagCloudContainerConfig.ImageSize.Width, _tagCloudContainerConfig.ImageSize.Height))
```
В других местах используется `var`, а не явный тип

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Core/Utils/ImageCreator.cs:
```diff
@@ -0,0 +1,30 @@
+
+namespace TagCloudContainer.Core.Utils;
+
+public class ImageCreator : IImageCreator
```
Кажется, этот класс не используется

dmitgaranin replied:
Все еще остался

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Core/Utils/TagCloudPlacer.cs:
```diff
@@ -0,0 +1,119 @@
+
+    private bool DoesItIntersect(Rectangle rectangle)
+    {
+        foreach (var putRectangle in _putRectangles)
```
Попробуй переписать на LINQ

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Core/Utils/WordValidator.cs:
```diff
@@ -0,0 +1,71 @@
+
+    private readonly ITagCloudContainerConfig _tagCloudContainerConfig;
+
+    public string Result
```
Можно сделать приватным

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Core/Utils/WordValidator.cs:
```diff
@@ -0,0 +1,71 @@
+
+public class WordValidator : IWordValidator 
+{
+    private StringBuilder _word;
```
Если к нему обратятся раньше вызова `ForWord`, будет NRE

dmitgaranin replied:
Осталось

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Core/Utils/WordValidator.cs:
```diff
@@ -0,0 +1,71 @@
+            .Where(line => !string.IsNullOrWhiteSpace(line));
+    }
+
+    private WordValidator ForWord(string word)
```
Не очень хорошее название метода

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Core/Utils/WordValidator.cs:
```diff
@@ -0,0 +1,71 @@
+        
+        return lines
+            .Where(line => !string.IsNullOrWhiteSpace(line))
+            .Select(w => ForWord(w)
```
Почему валидатор как-то изменяет входные данные? Он же должен лишь валидировать

dmitgaranin replied:
Осталось

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Core/Utils/WordValidator.cs:
```diff
@@ -0,0 +1,71 @@
+        _boringWords = File 
+            .ReadLines(filePath)
+            .Distinct()
+            .ToDictionary(w => w, w => 0);
```
Почему решил использовать словарь? Кажется, сейчас эта структура используется неоптимально. Давай ты придешь ко мне в ЛС и расскажешь о других структурах, которые могли бы здесь подойти лучше

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Core/Utils/WordsReader.cs:
```diff
@@ -0,0 +1,50 @@
+﻿using TagCloudContainer.Core;
```
В этом и других файлах есть неиспользуемые `usings`

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Forms/Register.cs:
```diff
@@ -0,0 +1,69 @@
+        catch (Exception e)
+        {
+            Console.WriteLine(e);
+            throw;
```
По условиям ТЗ не должно быть пробросов ошибок

dmitgaranin replied:
В текущей реализации проброс ошибок остался, но замаскирован в `GetValueOrThrow`

Нужно было возвращать все-все-все `Result` в корневой метод Main, который бы уже как-то реагировал на non-success ответ

1. Например, `TagCloudProvider.GetPreparedWords`:
`foreach (var word in arrangedWords.GetValueOrThrow())`
Лучшее решение - chain result'ов (пример описывал [ранее](https://github.com/kontur-courses/fp/pull/137#discussion_r1057074863)). 
И даже в твоем случае было бы достаточно `arrangedWords.Value`

2. `Registry.GetSettingsInstancesWithDefaultValues`
`tagCloudContainerConfig.WordsFilePath = wordsFilePath.IsSuccess ? wordsFilePath.GetValueOrThrow() : null;`
Тут тоже и делаешь проверку, и проверяешь еще раз через `GetValueOrThrow`
\+ еще кладешь `null` в `non-nullbale string`, что тоже плохо

В общем, я нашел таких 11 мест в решении

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Core/Utils/WordsArranger.cs:
```diff
@@ -0,0 +1,35 @@
+
+    public static Result<List<Word>> ArrangeWords(List<Word> words, bool random)
+    {
+        if (words == null)
```
В каких случаях этот параметр может быть `null`?

dmitgaranin commented on TagCloudContainer/TagCloudContainer.Forms/Validators/TagCloudFormConfigValidator.cs:
```diff
@@ -0,0 +1,27 @@
+    {
+        var systemFonts = new InstalledFontCollection().Families.Select(f => f.Name);
+        
+        if (tagCloudFormConfig == null)
```
В каких случаях он может быть `null`?

dmitgaranin replied:
Остались места в `TagCloudPlacerConfigValidator`

Luvr681 committed:
Разбил конфигурацию по разным файлам с отдельными ответственностями

Luvr681 committed:
1. Убрал проверку на null из конструктора и выбрасывание исключения всвязи с требованиями по задаче
2. Сделал проверку на то, существует ли путь к сохранению картинки или нет

Luvr681 committed:
Изменил имя валидатора строк

Luvr681 committed:
Выделил класс по работе с путями и вынес в него методы и конфига

Luvr681 committed:
Добавил валидацию для новых конфигов

Luvr681 committed:
Зарегистрировал недостающие зависимости и изменил регистрацию конфигов

Luvr681 committed:
1. Изменил код в связи с изменением валидации
2. Убрал из конструкторов проверки на null с выбрасыванием исключений

Luvr681 committed:
Изменил тесты в связи с изменением валидации конфигов и добавлением новых
